### ISSUE TITLE
On batchsize of Segment-level audio-visual contrastive pre-training

### ISSUE THREAD

**Issue by lzhangbj** [v-iashin/Synchformer#3]
Hi Vladimir,

Synchformer a great work! Congrats!

One question I have is the contrastive pretraining stage. The paper mentioned that negative pairs not only comes from (1) different segments in same video and (2) segments from different videos. The former represents synchronization contrast while the latter stands for semantics contrast. So there are essentially two objectives.

It seems that during training, you have computed the loss as a whole. I am wondering whether you have observed any conflicts on convergence of the two losses, i.e., training losses on sync contrast and semantics contrast, instead of both steadily decrease, are fighting each other.

I asked since appendix A mentioned that the batch size is only 2, and increasing batch size does not improve performance. This means, in one iteration, there are only two videos involved in semantics contrast. This is very different from conventional contrastive learning, where increasing batch size brings significant performance gains.

I am curious if you can kindly share any ideas/insights on the reason of the issue, and have more open discussions on this. 

Thank you.

**Comment by v-iashin** [comment]
hi thanks.

>The former represents synchronization contrast while the latter stands for semantics contrast.

We weren't conceptualizing these two parts of the loss this way. Although it does sound reasonable, I actually see them differently. It's a single loss that uses 50% easy samples (from another video) and 50% hard samples (segments from same video), aiming to optimize segment correspondence. My hypothesis was that including easy samples would make training smoother, but this, as well as other things, was largely untested due to the high compute needed for training S1. We haven’t tried using only hard samples (B=1). Also, I don’t view the first part as a sync loss, because our negatives aren’t very close to the positives in time—8 frames apart (0.32 sec)—unlike SyncNet, where they're separated by just 1 frame.

> I am wondering whether you have observed any conflicts on convergence of the two losses

I haven’t examined this specifically, but it’s an interesting question. Let me know if you find any insights. I’m also curious about experimenting with B=1 (segments from the same video).

> I asked since appendix A mentioned that the batch size is only 2

> This is very different from conventional contrastive learning, where increasing batch size brings significant performance gains.

Although we refer to it as batch size, it’s a bit different. We’re actually contrasting across segments, with 14 segments per video, so the effective "batch size" is 28 — still small compared to contrastive learning works that use batch sizes in the thousands. We were also limited by compute here. We did try to increase the batch size (number of videos) using a queue, but it didn’t improve performance, which is why we noted this in the appendix.

**Comment by lzhangbj** [comment]
Thanks for the quick response! 

I did try some similar experiments with our own encoders on contrastive learning with negative pairs coming from both same video and different videos. And one interesting finding is that, when printing out the "sync contrast" and "semantics contrast" losses separately, we find the two losses are actually fighting each other during training. Specifically, semantics contrast loss would, in a very long time, keep decreasing, while sync contrast loss would increase. This continues till some point where the sync loss would start decrease slowly, and contrast loss increase slowly meanwhile. We have tried some architecture changes and loss variations, but this phenomenon always happens.  This is the reason why I'm asking if you have observed similar things.

The observation made me think if the objectives of "semantics contrast" and "sync contrast" are inherently different. This is not the case for human perception, bc we can easily spot the sync and semantics differences at the same time. However, the current audio-video dataset are mostly noisy (I would say VERY noisy) in both audio-video semantics alignment and synchronization, and it turns out many trained models are actually using raw features instead of semantics features to distinguish synchronization. As a result, the contradict in sync contrast and align contrast just naturally happens during training.

And of course, the above is only my conjecture. I would like to hear from other opinions / solutions for sure.

Thanks for your sharing!

**Comment by v-iashin** [comment]
i see, thanks for sharing.

perhaps the reason why the 'semantics' loss is decreasing first and the 'sync' loss is not might be associated with the difficulty of doing the sync correspondence at earlier stages of training, justifying my hypothesis about hard/easy samples (curriculum learning basically). however, I don't think it is enough to justify it, though. 

to test if they are fighting each other (having one or the other is harmful for performance), I would compare these runs to the run with the 'sync' loss alone (bs=1), and observe if you could train it faster than with the two, and if the trained representation is actually better for the downstream task (or zero-shot). my bet would be that with B=1, it will work better, plus you could double the number of segments if you are training on something longer than vggsound or audioset to accommodate that (sorry I said 28 segments per video above, but I meant 14).

regarding the noise in the datasets. i agree with you. i wish we would have a better (cleaner) video dataset with strong av correspondence. with it, we could train better models for synchronization and other things like visually-guided sound generation.

i didn't get the raw vs semantic features point though. could you elaborate?

**Comment by lzhangbj** [comment]
Since the training videos are pretty noisy, and contrastive learning is applied to distinguish different segments in the same video, there is a chance that the model tend to give higher sync probabilities to those audio-videos where the peak/change in feature aligns more, even though they have nothing to do with sound generation. One case is camera movement, the other one is shot change. While for humans, even though there can be many (or very few) changes in pixel and audio space, we can identify the correspondence between the two, without being distracted by unrelated av changes. This is my understanding to "raw" feature changes and semantics changes.

Thanks for your suggestions and hypothesis! They are really helpful, and I believe there is many things to explore in this field.
